### PR TITLE
IH-28 Add api get coupon list for user

### PR DESCRIPTION
**Pull request by Sotatek-KimMai**

### CODE REVIEW

File: src/main/java/art/heredium/service/CouponUsageService.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+public class CouponUsageService {
+  private final CouponUsageRepository couponUsageRepository;
+
+  public List<CouponResponseDto> getCouponsWithUsageByAccountId(Long accountId) {
+    List<Coupon> coupons = couponUsageRepository.findDistinctCouponsByAccountId(accountId);
+    List<CouponResponseDto> responseDtos = new ArrayList<>();
+
+    for (Coupon coupon : coupons) {
+      List<CouponUsage> usedCouponUsages =
+          couponUsageRepository.findByAccountIdAndCouponId(accountId, coupon.getId()).stream()
```
Comment (Sotatek-ThoLe) [inline src/main/java/art/heredium/service/CouponUsageService.java:28]:
```suggestion
          couponUsageRepository.findUsedCoupons(accountId, coupon.getId())
```

File: src/main/java/art/heredium/service/CouponUsageService.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+              .filter(CouponUsage::getIsUsed)
+              .collect(Collectors.toList());
+
+      List<CouponUsage> unusedCouponUsages =
+          couponUsageRepository.findByAccountIdAndCouponId(accountId, coupon.getId()).stream()
```
Comment (Sotatek-ThoLe) [inline src/main/java/art/heredium/service/CouponUsageService.java:33]:
```suggestion
          couponUsageRepository.findUnusedCoupons(accountId, coupon.getId()).stream()
```

File: src/main/java/art/heredium/domain/coupon/model/dto/response/CouponResponseDto.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+@Getter
+@Setter
+public class CouponResponseDto {
+  private Long id;
+  private String name;
+  private String couponType;
+  private Integer discountPercent;
+  private String imageUrl;
+  private List<CouponUsageDto> usedCouponUsages;
+  private List<CouponUsageDto> unusedCouponUsages;
```
Comment (Sotatek-ThoLe) [inline src/main/java/art/heredium/domain/coupon/model/dto/response/CouponResponseDto.java:20]:
```suggestion
  private List<CouponUsageDto> usedCoupons;
  private List<CouponUsageDto> unusedCoupons;
```

File: src/main/java/art/heredium/service/CouponUsageService.java (hunk 1 above)
Comment (Sotatek-KimMai) [reply src/main/java/art/heredium/service/CouponUsageService.java]: I updated in 188a3a8b3c942b3d31a164d2b5876fba63f5d163

File: src/main/java/art/heredium/domain/coupon/model/dto/response/CouponResponseDto.java (hunk 3 above)
Comment (Sotatek-KimMai) [reply src/main/java/art/heredium/domain/coupon/model/dto/response/CouponResponseDto.java]: I updated in 188a3a8b3c942b3d31a164d2b5876fba63f5d163

File: src/main/java/art/heredium/domain/coupon/model/dto/response/CouponUsageDto.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  private Long id;
+  private LocalDateTime deliveredDate;
+  private LocalDateTime usedDate;
+  private LocalDateTime expirationDate;
+  private String uuid;
+
+  public CouponUsageDto(CouponUsage couponUsage) {
+    this.id = couponUsage.getId();
+    this.deliveredDate = couponUsage.getDeliveredDate();
+    this.usedDate = couponUsage.getUsedDate() != null ? couponUsage.getUsedDate() : null;
```
Comment (Sotatek-ThoLe) [inline src/main/java/art/heredium/domain/coupon/model/dto/response/CouponUsageDto.java:22]:
```suggestion
    this.usedDate = couponUsage.getUsedDate();
```

File: src/main/java/art/heredium/domain/coupon/repository/CouponUsageRepository.java [hunk 5]
```diff
@@ -1,9 +1,25 @@
...
+  @Query("SELECT DISTINCT cu.coupon FROM CouponUsage cu WHERE cu.account.id = :accountId")
+  List<Coupon> findDistinctCouponsByAccountId(@Param("accountId") Long accountId);
+
+  @Query(
+      "SELECT cu FROM CouponUsage cu WHERE cu.account.id = :accountId AND cu.coupon.id = :couponId AND cu.isUsed = :isUsed")
+  List<CouponUsage> findByAccountIdAndCouponIdAndIsUsed(
+      @Param("accountId") Long accountId,
+      @Param("couponId") Long couponId,
+      @Param("isUsed") boolean isUsed);
+}
```
Comment (Sotatek-ThoLe) [inline src/main/java/art/heredium/domain/coupon/repository/CouponUsageRepository.java:25]:
```suggestion
  List<CouponUsage> findByAccountIdAndCouponIdAndIsUsedTrue(
      @Param("accountId") Long accountId,
      @Param("couponId") Long couponId);
}

 List<CouponUsage> findByAccountIdAndCouponIdAndIsUsedFalse(
      @Param("accountId") Long accountId,
      @Param("couponId") Long couponId);
}
```
Thử 2 cái này xem. Nếu được thì đi theo hướng này, khỏi phải viết @Query, và cũng ko cần truyền thêm param isUsed

File: src/main/java/art/heredium/domain/coupon/model/dto/response/CouponUsageDto.java (hunk 4 above)
Comment (Sotatek-KimMai) [reply src/main/java/art/heredium/domain/coupon/model/dto/response/CouponUsageDto.java]: I updated in https://github.com/Incuvers-IT/heredium-BE/pull/9/commits/d285cb6158ae285e4a06c5e9bfff3eec0f9a967a

File: src/main/java/art/heredium/domain/coupon/repository/CouponUsageRepository.java (hunk 5 above)
Comment (Sotatek-KimMai) [reply src/main/java/art/heredium/domain/coupon/repository/CouponUsageRepository.java]: I updated in https://github.com/Incuvers-IT/heredium-BE/pull/9/commits/d285cb6158ae285e4a06c5e9bfff3eec0f9a967a

File: src/main/java/art/heredium/controller/user/UserCouponController.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+@RequestMapping("/api/user/coupons")
+public class UserCouponController {
+
+  private final CouponUsageService couponUsageService;
+
+  @GetMapping("/usage")
+  public ResponseEntity<List<CouponResponseDto>> getCouponsWithUsageByAccountId() {
+    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+    Long accountId = ((UserPrincipal) authentication.getPrincipal()).getId();
```
Comment (Sotatek-ThoLe) [inline src/main/java/art/heredium/controller/user/UserCouponController.java:29]: Đợi PR của Kim dc merge r call lại hàm này: https://github.com/Incuvers-IT/heredium-BE/pull/11/files#diff-55fa1109cbe4a30fa06498213d35d4502867199dde2f19e201f27b2498941f06R9

Comment (Sotatek-KimMai) [reply src/main/java/art/heredium/controller/user/UserCouponController.java]: I updated in https://github.com/Incuvers-IT/heredium-BE/pull/9/commits/f648d5882c64053e607c953296aaaf434a462505

**Review by Sotatek-ThoLe — APPROVED** [review]

**Pull request merged**